File: cat.pl
#!/usr/bin/perl

use strict;

sub cat {
	my $fh = @_[0];
	my $buffer = "x" x 2000;

	while (1) {
		my $n = sysread($fh, $buffer, 2000);
		if (!defined($n)) {
			die "Error reading: $!";
		}

		if ($n == 0) {
			last;
		}

		my $m = syswrite(STDOUT, $buffer, $n);
		if (!defined($m)) {
			die "Error writing: $!";
		}
	}
}

if ($#ARGV < 0) {
	open(my $fh, "<&STDIN") || die "Redirecting stdin: $!";
	cat($fh);
	close($fh) || die "Closing redirected stdin: $!";
} else {
	for (my $i = 0; $i <= $#ARGV; $i++) {
		open(my $fh, "<", $ARGV[$i]) || die "Opening $ARGV[$i]: $!";
		cat($fh);
		close($fh) || die "Closing $ARGV[$i]: $!";
	}
}

File: cat.go
package main

import "io"
import "os"
import "log"

func cat(fd *os.File) {
	buffer := make([]byte, 2000)

	for true {
		n, err := fd.Read(buffer)
		if err == io.EOF {
			break
		}
		if (err != nil) {
			log.Fatal(err)
		}

		var off = 0
		for off < n {
			m, err := os.Stdout.Write(buffer[off:n])
			if (err != nil) {
				log.Fatal(err)
			}

			off += m
		}
	}
}

func main() {
	if len(os.Args) < 2 {
		cat(os.Stdin)
	} else {
		for i := 1; i < len(os.Args); i++ {
			fd, err := os.Open(os.Args[i])
			if err != nil {
				log.Fatal(err)
			}

			cat(fd)

			err = fd.Close()
			if err != nil {
				log.Fatal(err)
			}
		}
	}
}
